File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
private :
    vector<int>dim={1,0,-1,0,1};
    void DFS(int r, int c, int&n,int&m,vector<vector<int>>&grid,bool&ans,int&sz)
    {
        if(r<0 or c<0 or r>=n or c>=m or grid[r][c]==0)
            return;
        
        if(r==0 or c==0 or r==n-1 or c==m-1)
            ans=false;
        
        grid[r][c]=0;
        sz++;
        
        for(int i=1;i<=4;i++)
            DFS(r+dim[i-1],c+dim[i],n,m,grid,ans,sz);
        
        return;        
    }
    
public:
    int numEnclaves(vector<vector<int>>& grid) {
        int n=grid.size(),m=grid[0].size(),count=0;
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==1)
                {
                    bool ans=true;
                    int sz=0;
                    DFS(i,j,n,m,grid,ans,sz);
                    
                    if(ans)
                        count+=sz;
                }
            }
        }
        
        return count;
        
    }
};

File: 1020-number-of-enclaves/1020-number-of-enclaves.java
class Solution {
    
    private int n,m,count;

    static final int[] dir = {1,0,-1,0,1};
    
    private boolean isValid(int i, int j,int[][] mat){
        
        if(i<0 || j<0 || i>=n || j>=m || mat[i][j] == 0) 
            return false;
        
        return true;
    }
    
    private void DFS(int i, int j, int[][] mat){
        
        if(!isValid(i,j,mat))
            return;
        
        mat[i][j] = 0;
        count++;
        
        for(int k=1;k<=4;k++){
            int x = dir[k-1] + i, y = dir[k] + j;
            DFS(x,y,mat);
        }
        
        return;
        
    }
    public int numEnclaves(int[][] grid) {
        
        n = grid.length;
        m = grid[0].length;
        count = 0;
        
        int oneCount = 0;
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j] == 1)
                    oneCount++;
            }
        }
        
        
        for(int i=0;i<m;i++){
            DFS(0,i,grid);
            DFS(n-1,i,grid);
        }
        
        for(int i=1;i<n;i++){
            DFS(i,0,grid);
            DFS(i,m-1,grid);
        }
        
        return oneCount - count;        
    }
}